File: src/java/src/com/zte/alphattt/Bench.java
package com.zte.alphattt;

import java.text.DecimalFormat;
import java.util.Arrays;

import com.zte.alphattt.game.Board;
import com.zte.alphattt.game.Mcts;
import com.zte.alphattt.game.Player;
import com.zte.alphattt.game.RandomPlayer;

public class Bench {
	Board board = new Board();
	Player player1;
	Player player2;
	
	public Bench(Player player1, Player player2) {
		this.player1 = player1;
		this.player2 = player2;
	}
	
	public void run(int games, int thinkMiliSecs){
		player1.setThinkMiliSecs(thinkMiliSecs);
		player2.setThinkMiliSecs(thinkMiliSecs);
		int i = 0;
		int player1Wins = 0;
		int player2Wins = 0;
		while(i++ < games){
			int winner = pkOneGame();
			if(winner == 1){
				player1Wins++;
			}else if(winner == 2){
				player2Wins++;
			}
			
			
			DecimalFormat df = new DecimalFormat("#.00"); 
			System.out.println("Total play " + i + " games, "
					+ this.player1.getClass().getSimpleName() + " wins " + player1Wins 
					+ "(" +  df.format((player1Wins*1.0 / i*1.0) * 100) + "%), "
					+ this.player2.getClass().getSimpleName() + " wins " + player2Wins 
					+ "(" + df.format((player2Wins*1.0 / i*1.0) * 100) + "%).");
		}
	}

	private int pkOneGame() {
		int[] state = board.start();
		int winner = -1;
		do{
			state = pkOneStep(this.player1, state);
			winner = board.winner(state);
			if(isEnd(winner)){
				return winner;
			}
			state = pkOneStep(this.player2, state);
			winner = board.winner(state);
			if(isEnd(winner)){
				return winner;
			}
		}while(true);
	}
	
	private int[] pkOneStep(Player player, int[] state){
		player.update(null, state);
		int[] move = player.getMove();
		state = board.next_state(state, move);
		//board.display(state);
		//System.out.println();
		return state;
	}
	
	private boolean isEnd(int winner){
		return winner == Board.PlAYER_1_WIN ||  winner == Board.PLAYER_2_WIN 
				|| winner == Board.DRAW;
	}
	
	public static void main(String[] args){
		Bench bench = new Bench(new Mcts(), new RandomPlayer());
		bench.run(100, 1000);
	}
	
}

File: src/python/playground/bench.py
# -*- coding: UTF-8 -*-
from client.board import Board
from player.mcts_ai import MctsAI
from player.mcts_ucb_ai import Mcts_UCB_AI

def run(player1_cls, player2_cls, count):
    print "Start battle!"
    cal_time, board = 0.1, Board()
    player1 = player1_cls(cal_time, board)
    player2 = player2_cls(cal_time, board)
    i = 0
    player1_win_count = 0;
    player2_win_count = 0;
    while i < count:
        winner = one_game(board, player1, player2)
        i += 1
        player1_win_count = player1_win_count + 1 if winner == player1.__class__ else player1_win_count
        player2_win_count = player2_win_count + 1 if winner == player2.__class__ else player2_win_count
        player1_win_rate = 100*(player1_win_count*1.0 / i)
        player2_win_rate = 100*(player2_win_count*1.0 / i)
        print "played %d games, "%i,  player1.__class__ \
            ," win %d(%.2f)," %(player1_win_count, player1_win_rate),\
            player2.__class__, " win %d(%.2f),"%(player2_win_count, player2_win_rate)   
            
def one_game(board, player1, player2):
    state = board.start()
    move = None
    winner = None
    while True:
        state, move, winner = one_step(state, move, board, player1, 1)
        if move == None:
            break
        state, move, winner = one_step(state, move, board, player2, 2)
        if move == None:
            break
    return winner
    
def one_step(state, move, board, player, winner):
    player.update(move, state)
    move = player.get_move()
    state = board.next_state(state, move)
    if board.winner(state) == winner:
        return state, None, player.__class__
    if board.winner(state) == Board.DRAW:
        return state, None, None
    return state, move, None

if __name__ == '__main__':
    run(MctsAI, Mcts_UCB_AI, 100)
